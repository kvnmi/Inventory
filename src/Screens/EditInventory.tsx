import React, { useState } from "react";
import { StyleSheet, Image, TouchableOpacity, Pressable } from "react-native";

import colors from "../components/colors";
import Form from "../components/Form";
import Field from "../components/FormFields";
import Screen from "../components/Screen";
import SubmitButton from "../components/SubmitButton";

import {
  fontPixel,
  heightPixel,
  normalise,
  pixelSizeHorizontal,
  pixelSizeVertical,
  widthPixel,
} from "../config/normalise";
import { HomeStackScreenProps } from "../Router/navTypes";
import { deleteInventoryItem, editInventory } from "../helpers/inventories";
import { IInventories } from "../interfaces/IInventory";
import { MaterialCommunityIcons } from "@expo/vector-icons";
import AppText from "../components/AppText";
import Modal from "../components/Modal";
import { validationSchema } from "../config/formValidationSchema";
import useAuth from "../Auth/useAuth";

export default function EditInventory({
  navigation,
  route,
}: HomeStackScreenProps<"EditListing">) {
  const { item } = route.params;
  const { user } = useAuth();

  const [failureModal, showFailureModal] = useState({
    active: false,
    message: "",
  });
  const [successModal, showSuccessModal] = useState<boolean>(false);
  const [deleteModal, showDeleteModal] = useState<boolean>(false);

  const closeModal = () => showFailureModal({ ...failureModal, active: false });

  async function handleUpdate(data: IInventories) {
    const response = await editInventory(data, user!.uid);
    if (!response.status) {
      showFailureModal({
        active: true,
        message: response.message,
      });
    } else {
      showSuccessModal(true);
    }
  }

  async function handleDelete() {
    await deleteInventoryItem(item.id, user!.uid);
    showDeleteModal(false);
    navigation.popToTop();
  }

  return (
    <Screen style={styles.container}>
      <Pressable style={styles.icon} onPress={navigation.goBack}>
        <MaterialCommunityIcons name="arrow-left" size={normalise(25)} />
      </Pressable>
      <Image style={styles.image} source={require("../assets/logo-red.png")} />
      <Form
        initialValues={{
          description: item.description,
          name: item.name,
          price: item.price,
          id: item.id,
          totalStock: item.totalStock,
        }}
        onSubmit={handleUpdate}
        validationSchema={validationSchema}
      >
        <Field
          fieldName="name"
          icon="card-text"
          maxLength={225}
          placeholder="Name"
          autoCapitalize="sentences"
          autoCorrect={true}
        />

        <Field
          fieldName="price"
          icon="cash"
          maxLength={8}
          placeholder="Price"
          autoCapitalize="none"
          autoCorrect={false}
          keyboardType="numeric"
        />
        <Field
          icon="counter"
          maxLength={8}
          placeholder="Stock"
          autoCapitalize="none"
          autoCorrect={false}
          keyboardType="numeric"
          fieldName={"totalStock"}
        />
        <Field
          icon="card-text"
          placeholder="Description"
          autoCapitalize="sentences"
          numberOfLines={3}
          autoCorrect={true}
          fieldName={"description"}
        />

        <SubmitButton title="Update Item" />
      </Form>

      <TouchableOpacity
        style={styles.deleteButton}
        onPress={() => showDeleteModal(true)}
      >
        <AppText style={styles.deleteText}>Delete Item</AppText>
      </TouchableOpacity>

      <Modal
        isActive={failureModal.active}
        onClose={closeModal}
        onPress={closeModal}
        confirmText=""
        deleteText="Close"
        message={failureModal.message}
        imgType="failure"
      />
      <Modal
        isActive={successModal}
        onClose={() => showSuccessModal(false)}
        onPress={navigation.goBack}
        confirmText="View Listing"
        deleteText="Continue"
        message={"Inventory listing added successfully"}
      />
      <Modal
        isActive={deleteModal}
        onClose={() => showDeleteModal(false)}
        onPress={handleDelete}
        confirmText="Continue"
        deleteText="Cancel"
        message={"Are you sure you want to delete this item?"}
      />
    </Screen>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: pixelSizeVertical(20),
    paddingHorizontal: pixelSizeHorizontal(15),
  },
  image: {
    height: heightPixel(80),
    width: widthPixel(80),
    marginBottom: pixelSizeHorizontal(35),
    alignSelf: "center",
  },
  deleteButton: {
    marginVertical: pixelSizeVertical(20),
    padding: normalise(20),
    alignSelf: "center",
  },
  deleteText: {
    color: colors.primary,
    fontSize: fontPixel(15),
    fontWeight: "700",
  },
  icon: {
    backgroundColor: colors.lightBlue,
    padding: normalise(5),
    borderRadius: 10,
    alignSelf: "flex-start",
    marginBottom: pixelSizeVertical(10),
  },
});
