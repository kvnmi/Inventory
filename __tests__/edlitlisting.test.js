import React from "react";
import renderer from "react-test-renderer";

import EditInventory from "../src/Screens/EditInventory";

describe("Edit Listing Screen", () => {
  const route = {
    params: {
      item: 1,
    },
  };
  const navigation = { navigate: jest.fn() };

  it("matches snapshot", () => {
    const tree = renderer
      .create(<EditInventory route={route} navigation={navigation} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
