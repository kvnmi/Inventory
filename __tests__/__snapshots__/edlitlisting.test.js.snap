// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Listing Screen matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    [
      "top",
    ]
  }
  style={
    [
      {
        "flex": 1,
      },
      {
        "flex": 1,
        "paddingHorizontal": 30,
        "paddingTop": 40,
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "flex-start",
        "backgroundColor": "#eef2ff",
        "borderRadius": 10,
        "marginBottom": 20,
        "padding": 10,
      }
    }
  >
    <Text />
  </View>
  <Image
    source={1}
    style={
      {
        "alignSelf": "center",
        "height": 132,
        "marginBottom": 70,
        "width": 160,
      }
    }
  />
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#e4e8e5",
          "borderRadius": 10,
          "flexDirection": "row",
          "marginVertical": 10,
          "padding": 8,
          "paddingLeft": 13,
        },
        {
          "width": "100%",
        },
      ]
    }
  >
    <Text />
    <TextInput
      autoCapitalize="sentences"
      autoCorrect={true}
      maxLength={225}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Name"
      placeholderTextColor="#6e6969"
      style={
        {
          "color": "black",
          "fontFamily": "Avenir",
          "fontSize": 28,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#e4e8e5",
          "borderRadius": 10,
          "flexDirection": "row",
          "marginVertical": 10,
          "padding": 8,
          "paddingLeft": 13,
        },
        {
          "width": "100%",
        },
      ]
    }
  >
    <Text />
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      keyboardType="numeric"
      maxLength={8}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Price"
      placeholderTextColor="#6e6969"
      style={
        {
          "color": "black",
          "fontFamily": "Avenir",
          "fontSize": 28,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#e4e8e5",
          "borderRadius": 10,
          "flexDirection": "row",
          "marginVertical": 10,
          "padding": 8,
          "paddingLeft": 13,
        },
        {
          "width": "100%",
        },
      ]
    }
  >
    <Text />
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      keyboardType="numeric"
      maxLength={8}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Stock"
      placeholderTextColor="#6e6969"
      style={
        {
          "color": "black",
          "fontFamily": "Avenir",
          "fontSize": 28,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#e4e8e5",
          "borderRadius": 10,
          "flexDirection": "row",
          "marginVertical": 10,
          "padding": 8,
          "paddingLeft": 13,
        },
        {
          "width": "100%",
        },
      ]
    }
  >
    <Text />
    <TextInput
      autoCapitalize="sentences"
      autoCorrect={true}
      numberOfLines={3}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Description"
      placeholderTextColor="#6e6969"
      style={
        {
          "color": "black",
          "fontFamily": "Avenir",
          "fontSize": 28,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fc5c65",
        "borderRadius": 15,
        "height": 100,
        "justifyContent": "center",
        "marginVertical": 20,
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 32,
        }
      }
    >
      Update Item
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "center",
        "marginVertical": 40,
        "opacity": 1,
        "padding": 40,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Avenir",
            "fontSize": 32,
          },
          {
            "color": "#fc5c65",
            "fontSize": 30,
            "fontWeight": "700",
          },
        ]
      }
    >
      Delete Item
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="none"
      deviceHeight={null}
      deviceWidth={null}
      hardwareAccelerated={false}
      hideModalContentWhileAnimating={false}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onRequestClose={[Function]}
      onSwipeComplete={[Function]}
      panResponderThreshold={4}
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      statusBarTranslucent={false}
      supportedOrientations={
        [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection={
        [
          "down",
        ]
      }
      swipeThreshold={100}
      transparent={true}
      visible={false}
    />
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="none"
      deviceHeight={null}
      deviceWidth={null}
      hardwareAccelerated={false}
      hideModalContentWhileAnimating={false}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onRequestClose={[Function]}
      onSwipeComplete={[Function]}
      panResponderThreshold={4}
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      statusBarTranslucent={false}
      supportedOrientations={
        [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection={
        [
          "down",
        ]
      }
      swipeThreshold={100}
      transparent={true}
      visible={false}
    />
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="none"
      deviceHeight={null}
      deviceWidth={null}
      hardwareAccelerated={false}
      hideModalContentWhileAnimating={false}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onRequestClose={[Function]}
      onSwipeComplete={[Function]}
      panResponderThreshold={4}
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      statusBarTranslucent={false}
      supportedOrientations={
        [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection={
        [
          "down",
        ]
      }
      swipeThreshold={100}
      transparent={true}
      visible={false}
    />
  </View>
</RNCSafeAreaView>
`;
